Fix compilation of temporal arithmetic in between filters

metamben opened the pull request:
Fixes #22531 for BigQuery and Mongo 5+.

metamben committed:
Fix compilation of temporal arithmetic in between for BigQuery

This fix only makes sure that relative temporal filters with offset generated by the
Metabase front end get compiled to a valid BigQuery query.

See #22531 for context.

metamben committed:
Fix compilation of temporal arithmetic in between for Mongo

camsaul commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:
```diff
@@ -217,6 +217,20 @@
+   :date      #{:day :week :month :quarter :year}
+   :time      #{:microsecond :millisecond :second :minute :hour}})
+
+(defmethod temporal-type :+
+  [[_ f [_interval _amount unit]]]
+  (when-let [field-type (temporal-type f)]
+    (if (and (= :timestamp field-type)
+             (not (contains? (temporal-type->supported-units :timestamp) unit)))
+      :datetime
+      field-type)))
```
This seems a little bit wonky it at least needs a comment explaining in detail why we're doing this. Otherwise if we come back and look at this in a few months it's going to be super confusing

camsaul replied:
At any rate this isn't how I'd go about doing this

This is the code that takes intervals and generates either `timestamp_add`, `datetime_add`, etc. is at

https://github.com/metabase/metabase/blob/master/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L627-L652

It seems like if you wanted to support doing something like adding a `:month` to a `:timestamp` you could just replace that `assert-addable-unit` stuff. Maybe something like this

```clj
(if (or (not add-fn)
        (not (addable-unit? t unit)))
  (recur (->temporal-type :datetime hsql-form))
  (hformat/to-sql (hsql/call add-fn hsql-form (interval amount unit))))
```

camsaul replied:
Oh nevermind. We're already actually doing almost the same thing you did here right here:

https://github.com/metabase/metabase/blob/master/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L662-L670

If you did something like `<timestamp> + [:interval -1 :month]` then it should give you `cast(<timestamp> AS datetime) + [:interval -1 :month]` which would ultimately compile to a `datetime_add()` and already have `temporal-type` of `:datetime`. So I'm not sure I understand what your change does. It should already work that way. I would suggest writing tests specifically for the `temporal-type` itself so it's clear what behavior you're changing here. (We also have tons of tests around `temporal-type` already, starting at https://github.com/metabase/metabase/blob/master/modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk/query_processor_test.clj#L274, and I would be doing anything to it without adding more tests.)

metamben replied:
The change fixes the bug by forcing the reconciliation of temporal units. The tweaked version of your test reproduces the problem. Are you suggesting I should omit the test cases that passed without the fix? (I kept it in order to increase my confidence that I don't break something here.)

camsaul replied:
Oh ok I think I understand what's happening. `reconcile-temporal-types` is getting called here

https://github.com/metabase/metabase/blob/master/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L604-L609

before we actually take the `:+` and call `add-interval-honeysql-form` to compile it, which happens here:

https://github.com/metabase/metabase/blob/master/src/metabase/driver/sql/query_processor.clj#L416-L424

So maybe the right to fix this is to move the call to `reconcile-temporal-types` to happen after the call to the parent compilation method instead of before that call. I don't think that would break anything but you'll have to try it out to be sure. Then you don't have to duplicate any logic anywhere, you just need to change

```clj
((get-method sql.qp/->honeysql [:sql filter-type])
 driver
 (reconcile-temporal-types clause))
```

to

```clj
(reconcile-temporal-types
 ((get-method sql.qp/->honeysql [:sql filter-type])
  driver
  clause))
```

That's what I'd try doing. Make sense?

> Are you suggesting I should omit the test cases that passed without the fix?

No.

metamben replied:
> That's what I'd try doing. Make sense?

Absolutely. I assumed there was some deep reason why it was not like that in the first place. The swap didn't work "out-of-the-box", but I can have a look if I can make it work or if I just did something wrong.

metamben replied:
@camsaul Just flipping the order of `sql.qp/->honeysql` and `reconcile-temporal-types` doesn't work. It breaks because it generates an `AddIntervalForm` which tries to add years to a timestamp. If I follow the code right, what happens is that
1. an `AddIntervalForm` w/o fixed temporal type is generated from the `[:relative-datetime -7 :year]` expression.
1. the field is converted to an `AddIntervalForm` with `:timestamp` as a fixed type,
1. `reconcile-temporal-types` forces all types to `:timestamp` (because that's the first temporal type it sees),
1. forcing the expression generated in step 1. to `:timestamp` generates no cast, but simply registers `:timestamp` as the type.
I found following the code quite difficult, so I might have misread how exactly we get this result, but I'm pretty sure that the problem is the invalid addition I mentioned.

What is even more difficult is judging what the best fix is, because I can only see what _is_ happening but cannot tell what _should_ be happing, what the intention was, etc.

What do you suggest? Go down this road further and fix the invalid `AddIntervalForm`?

camsaul replied:
> I found following the code quite difficult

Don't hesitate to rework it then if you have a way to make it clearer. We're a big 👍on constantly improving the code

> What do you suggest?

I played around with this REPL and figured it out. The issue is that the `AddIntervalForm` might be returning a `datetime` if it ends up using `datetime_add()` because the unit isn't allowed for `timestamp` (e.g. `month`), however `(temporal-type <AddIntervalForm>)` still returns `:timestamp` in this case because it hasn't done the switch to using `datetimes` yet (it doesn't happen until the call to `to-sql`).

So the fix here is to do that switch from a `timestamp` to a `datetime` right away instead of delaying it until `to-sql` compilation. That will mean `temporal-type` will return the right thing.

The second issue is that `->temporal-type` didn't do the right thing either. If you did

```clj
(->temporal-type :timestamp <AddIntervalForm>)
```

it would still produce something that's actually a `:datetime` if the unit wasn't allowed for `:timestamp`, since it attempted to update the underlying HoneySQL form rather than do any casting. The fix here is just to detect this specific case and add a cast to make sure we get the right type. 

I put together https://github.com/metabase/metabase/pull/23367 which demonstrates the changes I made to get things working correctly. There are also a couple of functions in there, `(x)` and `(y)` you can use for testing it from the REPL.

camsaul commented on modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj:
```diff
@@ -386,10 +387,16 @@
-                   [:>= field min-val]
-                   [:<= field max-val]]))
+  [[_ field min-val max-val :as expr]]
+  (if (= :+ (first field))
+    ;; filter with offset
+    (let [[_ field [_ amount unit]] field]
```
Probably not a good idea to assume every `[:+ ...]` has a `:field` and an `:interval` as its args at all. Here's a counter example

```clj
[:between [:+ [:field 1 nil] 2] ...]
```

or even

```clj
[:between [:+ [:interval -3 :day] [:field 2 nil]]]
```

camsaul replied:
Probably also a bad idea to assume `min-val` and `max-val` are integers. What if they're timestamps or something like that?

```clj
[:between
 [:+ [:field 1 {:temporal-unit :month}] [:interval -1 :day]]
 [:absolute-datetime #t "2022-06-13"]]

camsaul replied:
Either way this sort of transformation is not specific to MongoDB so it would make sense to do this as part of general query middleware so it applies to all databases. That would make it a lot easier to test as well.

metamben replied:
I completely agree that this is code is very brittle that's why I'm looking for advice where and how to do it more generally.
I'm a bit confused as to what you suggest. I had the impression that Mongo wasn't supporting an expression as the first argument of `:between`, that's why my tests failed. So I figured adding this minimal support won't make things worse. Are you suggesting we should make the scope larger and implement general support for expressions instead of just fields? And implement this transformation not just for Mongo but other drivers as well? (Postgres and MySQL don't seem to need, for example)).

camsaul replied:
Our MongoDB driver does support expressions as long as the DB is MongoDB 4.0 or above

https://github.com/metabase/metabase/blob/master/modules/drivers/mongo/src/metabase/driver/mongo.clj#L226-L231

camsaul replied:
You are right that the tests are failing because MongoDB isn't supporting expressions as arguments to `:between` (or other filter clauses), which should be allowed; however you don't need to "implement general support for expressions" since we already have that

camsaul replied:
After looking at this a bit more I think you just need to make sure we recursively call the appropriate compilation method on the args to the filters. I think `->rvalue` is what you want here. That's how the SQL QP version, `->honeysql`, does things. You just need to do that here too. e.g. recursively compile `field` (sort of a misnomer since it could be an expression or something else) so we get `{"$add" ...}` instead. Give that a try.

You could alternatively write a whole-query transformation to optimize that temporal arithmetic out entirely but that's more complicated and less general.

camsaul replied:
Also it looks like this bug probably affects other types of filter clauses as well such as `:<` or `:>=`. So you should write tests for those and make sure those get fixed as part of this PR

metamben replied:
I'm confused. Why doesn't it pass the tests then?

metamben replied:
Just transforming `field` with `->rvalue` doesn't currently work: it doesn't support translating `:interval`s. Now, I could try adding that, but the Mongo 4.2 docs say

> Adds numbers together or adds numbers and a date. If one of the arguments is a date, [$add](https://www.mongodb.com/docs/v4.2/reference/operator/aggregation/add/#exp._S_add) treats the other arguments as milliseconds to add to the date.

That makes translating `[:interval -3 :year]` on it's own impossible, the number of milliseconds we have to add depends on the date we are adding to. I have no idea at all how to translate that to milliseconds if we are adding to a field.

camsaul replied:
I think using `$dateAdd` is fine. We can throw an Exception if the Database version isn't 5.0+ advising people that we only support it for Mongo 5 or above. It's a tiny edge case we didn't know about until now, so I think that's an acceptable solution. 

You'll just have to figure out how to only run that test for Mongo if we're testing against 5 or above... `metabase.driver/describe-database` is supposed to return `:version` so you can write some sort of logic that only runs the test for Mongo if version is >= 5.0. A little hacky but we don't have a better way of doing that at this point in time and since it's a one-off I think it would be fine as long as you write down why we're doing it.

metamben replied:
@camsaul At what point should the DB version be checked? Is it available in some context (I think I've seen some kind of store "store") during compilation? Can I delay it until the point when we get to compiling the date arithmetic?

camsaul replied:
Yeah I think doing the check when compiling the arithmetic is fine. `(qp.store/db)` will give you the current Database that we're running the query against so you can pass that in to `driver/details` or whatever you're using to get the version

camsaul commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:
```diff
@@ -217,6 +217,20 @@
+(defmethod temporal-type :+
+  [[_ f [_interval _amount unit]]]
+  (when-let [field-type (temporal-type f)]
+    (if (and (= :timestamp field-type)
```
Nitpick but I'd read this as `:timestamp equals field-type` which is a [Yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions)

metamben replied:
Will reverse.

camsaul commented on test/metabase/query_processor_test/filter_test.clj:
```diff
@@ -109,6 +109,30 @@
+    ;; we also want to test this against MongoDB but [[mt/normal-drivers-with-feature]] would normally not include that
+    ;; since MongoDB only supports expressions if version is 4.0 or above and [[mt/normal-drivers-with-feature]]
+    ;; currently uses [[driver/supports?]] rather than [[driver/database-supports?]] (TODO FIXME)
+    (mt/with-clock #t "2022-06-10T16:12-08:00[US/Pacific]"
```
`with-clock` does not affect the clocks of the Databases we connect to so it's not going to affect relative datetimes or anything like that. If this test passes now it will start breaking at some point in the future

metamben replied:
Will reverse.

camsaul replied:
If you want to make sure that we actually get the correct rows you can take a look at the tests in https://github.com/metabase/metabase/blob/master/test/metabase/query_processor_test/date_bucketing_test.clj which actually create a new dataset on the fly to run tests against so we can verify that relative datetime things actually return the correct rows. You can use those without using `with-clock`

camsaul commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:
```diff
@@ -217,6 +217,20 @@
+   :time      #{:microsecond :millisecond :second :minute :hour}})
+
+(defmethod temporal-type :+
+  [[_ f [_interval _amount unit]]]
```
Same as with the Mongo stuff but you can add more to a column than just an `:interval`

metamben committed:
Add more general fix based on Cam's solution (#23367)

metamben committed:
Add more general fix for Mongo 5

Mongo 4 doesn't support $dateAdd so the generated filters result in an exception.

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj:
```diff
@@ -363,8 +368,15 @@
-(defmethod ->rvalue :- [[_ & args]] {"$subtract" (mapv ->rvalue args)})
+(defmethod ->rvalue :+ [[_ & [lhs rhs :as args]]]
+  (cond
+    (= (first lhs) :interval) (date-op "$dateAdd" rhs lhs)
```
Is this order valid in MBQL?

camsaul replied:
Conceptually it seems like it should be valid, but that schema is probably overly strict. I would still try to support it here so when we fix the schema we won't have to go back and fix this too.

Dumb question: wouldn't it make more sense to just implement `->rvalue` for `:interval`? Because otherwise this wouldn't fix `[:- <whatever> [:interval ...]]` forms. Only `:+`

camsaul replied:
Oh nevermind I guess that wouldn't work because we have to use `$dateAdd` instead of `$add`. Well either way it would still be good to try to support this for `:-` forms as well. Better to fix it while we're poking around this part of the codebase instead of it getting opened as an issue in the future

camsaul commented on test/metabase/query_processor_test/filter_test.clj:
```diff
@@ -109,6 +110,46 @@
+    ;; we also want to test this against MongoDB but [[mt/normal-drivers-with-feature]] would normally not include that
+    ;; since MongoDB only supports expressions if version is 4.0 or above and [[mt/normal-drivers-with-feature]]
+    ;; currently uses [[driver/supports?]] rather than [[driver/database-supports?]] (TODO FIXME)
+    (mt/test-drivers (conj (mt/normal-drivers-with-feature :expressions) :mongo)
+      (mt/dataset attempted-murders
```
Actually MongoDB 3.x was EOL'ed over a year ago so at this point it's probably safe to say we don't support MongoDB 3.x anymore. We don't test against it anyway -- we test against 4.0 and 5.0 IIRC. You could actually just replace that `database-supports?` method for `[:mongo :expressions]` with a `supports?` method and then we won't need this special case code here if you want. That would give us the added benefit of having other `:expressions` code run against Mongo without us needing special case code here. Might catch some more bugs

metamben replied:
That seems to provoke quite some test failures so I prefer to do it separately, not in this PR.

camsaul commented on modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj:
```diff
@@ -363,8 +371,15 @@
 
-(defmethod ->rvalue :+ [[_ & args]] {"$add" (mapv ->rvalue args)})
-(defmethod ->rvalue :- [[_ & args]] {"$subtract" (mapv ->rvalue args)})
+(defmethod ->rvalue :+ [[_ & [lhs rhs :as args]]]
+  (cond
+    (interval? lhs) (date-op "$dateAdd" rhs lhs)
+    (interval? rhs) (date-op "$dateAdd" lhs rhs)
+    :else {"$add" (mapv ->rvalue args)}))
+(defmethod ->rvalue :- [[_ & [lhs rhs :as args]]]
+  (if (interval? rhs)
+    (date-op "$dateSubtract" lhs rhs)
+    {"$subtract" (mapv ->rvalue args)}))
```
```suggestion

(defmethod ->rvalue :+ [[_ & [lhs rhs :as args]]]
  (cond
    (interval? lhs) (date-op "$dateAdd" rhs lhs)
    (interval? rhs) (date-op "$dateAdd" lhs rhs)
    :else {"$add" (mapv ->rvalue args)}))
    
(defmethod ->rvalue :- [[_ & [lhs rhs :as args]]]
  (if (interval? rhs)
    (date-op "$dateSubtract" lhs rhs)
    {"$subtract" (mapv ->rvalue args)}))
    
```

Please add some newlines here, having no lines between them stops making sense when the forms are more than one line

camsaul reviewed (commented):
Overall I think this is like 99% of the way there but I have a few minor requests:

1. Can we add support for `[:- <interval> <whatever>]` for MongoDB as well? Seems weird to support it for `:+` but not `:-`.
2. It would be nice to add a `:-` to your test as well if you can
3. It would be good to look and see if you can write this in a way that we can support arithmetic with multiple intervals. Technically `:+` and `:-` are variadic in MBQL so theoretically you _should_ be able to do something like `[:+ <field> <interval> <interval>]`. I doubt this is going to work out of the box tho. I don't want the scope of this bugfix to grow out of control so see if you can fix that easily as part of this PR. If you can't then please open a GH issue so we don't forget to fix it at some point in the future. It would be nice to throw some sort of "this is currently unsupported" error -- you could do that as part of the MBQL schema maybe -- and add a test for that -- but I would not say that's something you need to do as part of this PR, unless you think you can do it without too much hassle.
4. Please open a GH issue that `[:+ <interval> <field>]` (or `:-`) throws a schema validation error even tho theoretically it seems like something we should support

camsaul reviewed (approved):
Going to go ahead and give this a 👍 since it looks good so you're not blocked on asking me to do another review. Everything looks pretty close so just make those last few minor tweaks, get it green, and then hit the merge button

metamben commented:
@camsaul Here are my answers. (I'm afraid there is still some confusion as to what exactly is to be done.)

> 1. Can we add support for `[:- <interval> <whatever>]` for MongoDB as well? Seems weird to support it for `:+` but not `:-`.

I don't think subtracting from intervals is supported and I don't think it's very useful either. I don't find the lack of symmetry weird, as `-` is not commutative.

> 2. It would be nice to add a `:-` to your test as well if you can

I had that in my test, but it seemed to me that it's not supported either. For example, `sql.qp/add-interval-honeysql-form` exists, but `sql.qp/subtract-interval-honeysql-form` (or something equivalent) doesn't.

> 3. It would be good to look and see if you can write this in a way that we can support arithmetic with multiple intervals. 

Will look into this.

> 4. Please open a GH issue that `[:+ <interval> <field>]` (or `:-`) throws a schema validation error even tho theoretically it seems like something we should support

OK, but I can't attach obvious semantics to `[:- <interval> <field>]`. Interval arithmetic is not commonplace, AFAIK.

I probably can't merge just yet because oracle, presto, redshift, snowflake and sparksql are failing and I haven't look into those failures yet. :(

metamben committed:
Support adding field to interval too

Time intervals were not allowed in the first place of an addition.

metamben committed:
Support temporal arithmetic with more than two operands for Mongo

metamben committed:
Restrict tests to timestamp column aware databases

metamben commented:
@flamber This fix doesn't work with Mongo versions before 5. How much effort should we spend on supporting Mongo 4?

flamber commented:
@metamben We didn't implement Custom Columns on v4 for some reason, guess limitation but I don't know - #17117.
The [EOL](https://www.mongodb.com/support-policy/lifecycles) for 4.x is now only less than two years away, then it makes even less sense to work on that version unless it's easy.

metamben committed:
Add mixed temporal arithmetic test

metamben commented:
> @metamben We didn't implement Custom Columns on v4 for some reason, guess limitation but I don't know - #17117. The [EOL](https://www.mongodb.com/support-policy/lifecycles) for 4.x is now only less than two years away, then it makes even less sense to work on that version unless it's easy.

I think the change is not trivial, I can see at least two places where special handling would have to be introduced for version 4. I suggest we don't implement it until people start complaining about it and we can assess how important it is to fix this problem.

metamben commented on src/metabase/util.clj:
```diff
@@ -983,3 +983,14 @@
   []
   (nano-id))
+
+(defn pick-first
```
Is this the right place for this function? Are the trade-offs OK?

dpsutton replied:
looks fine to me.

metamben committed:
Mention followup ticket in comment

metamben committed:
Fix comment

metamben committed:
Refer GH issue in the TODO comment
